README.md
# Benchmark

A Clojure library designed to ... well, that part is up to you.

## Usage

```clojure
(def n (doall (range 1 1e6)))
(def nd (double-array n))
```

## Benchmark Results

### Count Calculations
| Function       | Execution Time | Improvement (%) | Speed Factor   |
|----------------|----------------|-----------------|----------------|
| `count` on `n` | 37.41 ms       | -               | -              |
| `count` on `nd`| 58.47 ns       | 99.84%          | x639.63        |

### Average Calculations
| Function                     | Execution Time | Improvement (%) | Speed Factor   |
|------------------------------|----------------|-----------------|----------------|
| `avg-reduce` on `n`          | 46.38 ms       | -               | -              |
| `avg-loop-rec` on `n`        | 22.37 ms       | 51.80%          | x2.07          |
| `avg-areduce` on `nd`        | 733.58 µs      | 98.42%          | x63.21         |
| `avg-loop-rec-array` on `nd` | 728.14 µs      | 98.43%          | x63.68         |

### Standard Deviation Calculations
| Function                       | Execution Time | Improvement (%) | Speed Factor   |
|--------------------------------|----------------|-----------------|----------------|
| `std-dev-reduce` on `n`        | 89.61 ms       | -               | -              |
| `std-dev-loop-rec` on `n`      | 66.46 ms       | 25.84%          | x1.35          |
| `std-dev-areduce` on `nd`      | 1.48 ms        | 98.35%          | x60.55         |
| `std-dev-loop-rec-array` on `nd` | 1.49 ms     | 98.34%          | x60.14         |

### Dot Product Calculations
| Function                           | Execution Time | Improvement (%) | Speed Factor   |
|------------------------------------|----------------|-----------------|----------------|
| `dot-product-loop-rec` on `nd`     | 30.56 ms       | -               | -              |
| `dot-product-loop-rec-array` on `nd` | 751.90 µs  | 97.54%          | x40.66         |



## License

Copyright © 2023 Đorđe Novaković

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-12-15
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-12-15
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/benchmark/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/benchmark/compare/0.1.0...0.1.1


src/benchmark/mesurements.clj
(ns benchmark.mesurements
  (:require [criterium.core :refer :all]
            [benchmark.core :refer :all]))

(def n (doall (range 1 1e6)))
(def nd (double-array n))
(quick-bench (count n))                                     ;  37,408835 ms
(quick-bench (count nd))                                    ;  58,474716 ns

; Average benchmark

(avg-reduce n)
(quick-bench (avg-reduce n))                                ;  46,381846 ms

(avg-loop-rec n)
(quick-bench (avg-loop-rec n))                              ;  22,371506 ms

(avg-areduce nd)
(quick-bench (avg-areduce nd))                              ; 733,582998 µs

(avg-loop-rec-array nd)
(quick-bench (avg-loop-rec-array nd))                       ; 728,138070 µs

; Standard deviation benchmark

(std-dev-reduce n)
(quick-bench (std-dev-reduce n))                            ;  89,611946 ms

(std-dev-loop-rec n)
(quick-bench (std-dev-loop-rec n))                          ;  66,459096 ms

(std-dev-areduce nd)
(quick-bench (std-dev-areduce nd))                          ;   1,483616 ms

(std-dev-loop-rec-array nd)
(quick-bench (std-dev-loop-rec-array nd))                   ;   1,485625 ms

; dot product benchmark

(dot-product-loop-rec nd nd)
(quick-bench (dot-product-loop-rec nd nd))                  ;  30,556188 ms

(dot-product-loop-rec-array nd nd)
(quick-bench (dot-product-loop-rec-array nd nd))            ; 751,898635 µs



src/benchmark/core.clj
(ns benchmark.core
  (:require [clojure.math :refer :all]))

(set! *warn-on-reflection* true)
(set! *unchecked-math* :warn-on-boxed)
(def n (doall (range 1 1e6)))
(def nd (double-array n))

; Average benchmark

;  46,381846 ms
(defn avg-reduce ^double [n]
  (/ ^double (reduce + n) (count n)))
(avg-reduce n)

;  22,371506 ms
(defn avg-loop-rec [n]
  (loop [sum 0.0 i 0 s (seq n)]
    (if s
      (recur (+ sum ^double (first s)) (inc i) (next s))
      (/ sum i))))
(avg-loop-rec n)

; 733,582998 µs
(defn avg-areduce [^doubles n]
  (/ (double (areduce n i acc 0.0 (+ acc (aget n i)))) (count n)))

(defn avg-areduce [^doubles n]
  (/ ^double (areduce n i acc 0.0 (+ acc (aget n i))) (count n)))
(avg-areduce nd)

; 728,138070 µs
(defn avg-loop-rec-array [^doubles n]
  (loop [sum 0.0 i 0 size (count n)]
    (if (< i size)
      (recur (+ sum ^double (aget n i)) (inc i) size)
      (/ sum i))))
(avg-loop-rec-array nd)

;(defn avg-loop-rec-array2 [^doubles n]
;  (loop [sum 0.0 i 0 s (seq n)]
;    (if s
;      (recur (+ sum ^double (first s)) (inc i) (next s))
;      (/ sum i))))
;
;(avg-loop-rec-array2 nd)

; Standard deviation benchmark

; 89,611946 ms
(defn std-dev-reduce ^double [n]
  (let [avg (avg-reduce n)]
    (sqrt (/ (reduce (fn [acc e]
                       (+ acc (pow (- avg e) 2))) 0.0 n) (count n)))))
(std-dev-reduce n)

; 66,459096 ms
(defn std-dev-loop-rec ^double [n]
  (let [avg (avg-reduce n)]
    (loop [sum 0.0
           i 0
           n (seq n)]
      (if n
        (recur (+ sum (pow (- avg (first n)) 2)) (inc i) (next n))
        (sqrt (/ sum i))))))
(std-dev-loop-rec n)

;  1,483616 ms
(defn std-dev-areduce ^double [^doubles n]
  (let [avg (avg-areduce n)]
    (sqrt (/ (areduce n idx ret 0.0 (+ ret (pow (- avg (aget n idx)) 2)))
             (count n)))))
(std-dev-areduce nd)

;  1,485625 ms
(defn std-dev-loop-rec-array ^double [^doubles n]
  (let [avg (avg-areduce n)]
    (loop [sum 0.0
           i 0
           size (count n)]
      (if (< i size)
        (recur (+ sum (pow (- avg (aget n i)) 2)) (inc i) size )
        (sqrt (/ sum size))))))
(std-dev-loop-rec-array nd)

; dot product

(defn dot-product-loop-rec ^double [n m]
  (loop [sum 0.0
         i 0
         s1 (seq n)
         s2 (seq m)]
    (if s1
      (recur (+ sum (* (first s1) (first s2))) (inc i) (next s1) (next s2))
      sum)))
(dot-product-loop-rec nd nd)

;  4,869953 ms
(defn dot-product-loop-rec-array ^double [^doubles n ^doubles m]
  (loop [sum 0.0
         i 0
         size (count n)]
    (if (< i size)
      (recur (+ sum (* (aget n i) (aget m i))) (inc i) size )
      sum)))
(dot-product-loop-rec-array nd nd)









































